wagtail_streamfield_migration_toolkit/autodetect/comparers.py

```diff
@@ -1,156 +1,269 @@
+        # old block is a StreamBlock and compared block is a CharBlock, it is fair to say that we
+        # can be certain it is not a renamed block.
+        # TODO consider eg. like CharBlock, TextBlock. In that case do we give this a weight and
+        # return a non zero score for blocks like char and text that may have compatible content?
+        if not cls.compare_types(old_path, new_path):
+            return 0
+
+        name_similarity = cls.compare_names(old_name, new_name)
+        arg_similarity = cls.compare_args(old_args, new_args)
+        kwarg_similarity = cls.compare_kwargs(old_kwargs, new_kwargs)
 
-        # TODO this might not be a black and white decision for all cases
-        # However for cases like where old block is a StreamBlock and compared block is a CharBlock,
-        # it is fair to say that we can be certain it is not a renamed block.
-        # TODO consider eg. like CharBlock, TextBlock
-        if not cls.compare_types(old_def, new_def):
-            return False
-
-        # name_similarity = cls.compare_names(old_name, new_name) # TODO
+        return (
+            arg_similarity * cls.arg_weight
+            + name_similarity * cls.name_weight
+            + kwarg_similarity * cls.kwarg_weight
+        ) / (cls.arg_weight + cls.name_weight + cls.kwarg_weight)
 
-        arg_similarity = cls.compare_args(old_def, new_def)
+    @classmethod
+    def compare(cls, old_def, old_name, new_def, new_name):
 
-        child_similarity = cls.compare_children(old_def, new_def)
+        if not cls.compare_types_initial(old_def, new_def):
+            return 0
+
+        # TODO it might be best to add some separate tests for the hashable_deep_deconstruct method
+        # itself
+        old_path, old_args, old_kwargs = cls.hashable_deep_deconstruct(old_def)
+        new_path, new_args, new_kwargs = cls.hashable_deep_deconstruct(new_def)
+
+        # - For structural blocks, args are a list of children, and kwargs contain block options
+        # like label, icon etc.
+        # - For other blocks, args includes any positional arguments and kwargs contains the block
+        # options. For example, SnippetChooser block would have the snippet class as a positional
+        # argument. Most basic blocks like CharBlock have no args it seems.
+
+        return cls._compare(
+            old_name=old_name,
+            old_path=old_path,
+            old_args=old_args,
+            old_kwargs=old_kwargs,
+            new_name=new_name,
+            new_path=new_path,
+            new_args=new_args,
+            new_kwargs=new_kwargs,
+        )
 
-        return (
-            arg_similarity * cls.arg_weight + child_similarity * cls.child_weight
-        ) / (cls.arg_weight + cls.child_weight)
+    @staticmethod
+    def compare_types_initial(old_def, new_def):
+        old_path, _, _ = old_def.deconstruct()
+        new_path, _, _ = new_def.deconstruct()
+        return old_path == new_path
```

Comment:
It was actually meant to avoid calling hashable_deep_deconstruct and call _compare for block types which would anyway not be compatible, but since we're probably going to be comparing every block with at least one block, perhaps we might remove this method

Reply:
removed initial type comparison